File: 3.Longest Common Substring/LCS.cpp
#include <bits/stdc++.h>
using namespace std;

int LCS_Iterative(string & s1, string& s2) {
    int res = 0;
    int n = s1.length(), m = s2.length();
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            int curr = 0;
            while ((i + curr) < n && (j + curr) < m && s1[i + curr] == s2[j + curr]) 
                curr++;
            res = max(res, curr);
        }
    }
    return res;
}

int LCS_Solve(const string& s1, const string& s2, int n, int m) {
    if (n == 0 || m == 0 || s1[n - 1] != s2[m - 1])
        return 0;
    return 1 + LCS_Solve(s1, s2, n - 1, m - 1);
}

int LCS_Recursion(const string& s1, const string& s2) {
    int res = 0;
    for (int i = 1; i <= s1.size(); i++) 
        for (int j = 1; j <= s2.size(); j++) 
            res = max(res, LCS_Solve(s1, s2, i, j));
    return res;
}

int LCS_Tabulation(const string& s1, const string& s2) {
    int res = 0;
    int m = s1.length(), n = s2.length();
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));    
    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (s1[i - 1] == s2[j - 1]) {
                dp[i][j] = 1 + dp[i - 1][j - 1];
                res = max(res, dp[i][j]);
            }
            else 
                dp[i][j] = 0;
        }
    }
    return res;
}

int LCS_Tabulation_SpaceOptimized(const string& s1, const string& s2) {
    int res = 0;
    int n = s1.length(), m = s2.length();
    vector<int> prev(n + 1, 0), cur(m + 1, 0);
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            if (s1[i - 1] == s2[j - 1]) {
                cur[j] = prev[j - 1] + 1;
                res = max(res, cur[j]);
            }
            else 
                cur[j] = 0;
        }
        prev = cur;
    }
    return res;
}

string LCS_Tabulation_Print(const string& s1, const string& s2) {
    int m = s1.length(), n = s2.length();
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
    int maxLen = 0, endIndex = -1; 

    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (s1[i - 1] == s2[j - 1]) {
                dp[i][j] = 1 + dp[i - 1][j - 1];
                if (dp[i][j] > maxLen) {
                    maxLen = dp[i][j];
                    endIndex = i - 1; 
                }
            } 
            else 
                dp[i][j] = 0; 
        }
    }

    string lcs = "";
    if (maxLen > 0) 
        lcs = s1.substr(endIndex - maxLen + 1, maxLen);

    cout << "Longest Common Substring: " << lcs << " || and length is: " << maxLen << endl;

    return lcs;
}

string LCS_Tabulation_SpaceOptimized_Print(const string& s1, const string& s2) {
    int n = s1.length(), m = s2.length();
    vector<int> prev(m + 1, 0), cur(m + 1, 0);
    int maxLen = 0, endIndex = -1; 

    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            if (s1[i - 1] == s2[j - 1]) {
                cur[j] = prev[j - 1] + 1;
                if (cur[j] > maxLen) {
                    maxLen = cur[j];
                    endIndex = i - 1; 
                }
            } 
            else 
                cur[j] = 0; 
        }
        prev = cur;
    }

    string lcs = "";
    if (maxLen > 0) 
        lcs = s1.substr(endIndex - maxLen + 1, maxLen);

    cout << "Longest Common Substring: " << lcs << " || and length is: " << maxLen << endl;

    return lcs;
}


int main() {
    string s1 = "geeksforgeeks";
    string s2 = "practicewritegeekscourses";
    
    cout << "Longest Common Substring Iterative: " << LCS_Iterative(s1, s2) << endl;
    cout << "Longest Common Substring Recursion: " << LCS_Recursion(s1, s2) << endl;
    cout << "Longest Common Substring Tabulation: " << LCS_Tabulation(s1, s2) << endl;
    cout << "Longest Common Substring Tabulation_SpaceOptimized: " << LCS_Tabulation_SpaceOptimized(s1, s2) << endl;

    LCS_Tabulation_Print(s1, s2);
    LCS_Tabulation_SpaceOptimized_Print(s1, s2);

    return 0;
}

File: 10. Dynamic Programming/03. Longest Common Substring/LCS.go
package main

import (
	"fmt"
)

// 1. Iterative Brute Force
func LCSIterative(s1, s2 string) int {
	res := 0
	n, m := len(s1), len(s2)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			curr := 0
			for (i+curr) < n && (j+curr) < m && s1[i+curr] == s2[j+curr] {
				curr++
			}
			if curr > res {
				res = curr
			}
		}
	}
	return res
}

// Helper for Recursion
func LCSSolve(s1, s2 string, n, m int) int {
	if n == 0 || m == 0 || s1[n-1] != s2[m-1] {
		return 0
	}
	return 1 + LCSSolve(s1, s2, n-1, m-1)
}

// 2. Recursive
func LCSRecursion(s1, s2 string) int {
	res := 0
	for i := 1; i <= len(s1); i++ {
		for j := 1; j <= len(s2); j++ {
			temp := LCSSolve(s1, s2, i, j)
			if temp > res {
				res = temp
			}
		}
	}
	return res
}

// 3. Tabulation
func LCSTabulation(s1, s2 string) int {
	res := 0
	n, m := len(s1), len(s2)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
	}
	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if s1[i-1] == s2[j-1] {
				dp[i][j] = 1 + dp[i-1][j-1]
				if dp[i][j] > res {
					res = dp[i][j]
				}
			}
		}
	}
	return res
}

// 4. Space Optimized
func LCSTabulationSpaceOptimized(s1, s2 string) int {
	n, m := len(s1), len(s2)
	prev := make([]int, m+1)
	cur := make([]int, m+1)
	res := 0

	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if s1[i-1] == s2[j-1] {
				cur[j] = prev[j-1] + 1
				if cur[j] > res {
					res = cur[j]
				}
			} else {
				cur[j] = 0
			}
		}
		copy(prev, cur)
	}
	return res
}

// 5. Tabulation with Print
func LCSTabulationPrint(s1, s2 string) string {
	n, m := len(s1), len(s2)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, m+1)
	}

	maxLen, endIndex := 0, -1
	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if s1[i-1] == s2[j-1] {
				dp[i][j] = 1 + dp[i-1][j-1]
				if dp[i][j] > maxLen {
					maxLen = dp[i][j]
					endIndex = i - 1
				}
			}
		}
	}

	if maxLen > 0 {
		return s1[endIndex-maxLen+1 : endIndex+1]
	}
	return ""
}

// 6. Space Optimized Print
func LCSTabulationSpaceOptimizedPrint(s1, s2 string) string {
	n, m := len(s1), len(s2)
	prev := make([]int, m+1)
	cur := make([]int, m+1)
	maxLen, endIndex := 0, -1

	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if s1[i-1] == s2[j-1] {
				cur[j] = prev[j-1] + 1
				if cur[j] > maxLen {
					maxLen = cur[j]
					endIndex = i - 1
				}
			} else {
				cur[j] = 0
			}
		}
		copy(prev, cur)
	}

	if maxLen > 0 {
		return s1[endIndex-maxLen+1 : endIndex+1]
	}
	return ""
}

// Main Function
func main() {
	s1 := "geeksforgeeks"
	s2 := "practicewritegeekscourses"

	fmt.Println("LCS Iterative:", LCSIterative(s1, s2))
	fmt.Println("LCS Recursion:", LCSRecursion(s1, s2))
	fmt.Println("LCS Tabulation:", LCSTabulation(s1, s2))
	fmt.Println("LCS Tabulation Space Optimized:", LCSTabulationSpaceOptimized(s1, s2))

	fmt.Println("LCS Tabulation Print:", LCSTabulationPrint(s1, s2))
	fmt.Println("LCS Tabulation Space Optimized Print:", LCSTabulationSpaceOptimizedPrint(s1, s2))
}
